### PR TITLE
unify build finding logic

### PR DESCRIPTION
**Pull request by grosser**
@jonmoter @ragurney 

WIP ... but where I think we need to go

### CODE REVIEW

File: lib/samson/build_finder.rb [hunk 1]
```diff
@@ -26,51 +26,55 @@ def cancelled!
...
+      needed = (@images ? @images : @job.project.dockerfile_list.map { |d| [d, @job.project.docker_image(dockerfile)] })
+      build_disabled = @job.project.docker_image_building_disabled?
+      all = []
+
+      wait_for_build_creation do |last_try|
+        needed.delete_if do |selector|
+          found = detect_build_by_selector!(possible_builds, *selector, fail: (last_try && build_disabled))
+          if found
+            all << found
+          elsif last_try && !build_disabled
```
Comment (jonmoter) [inline lib/samson/build_finder.rb]: Kindof duplicated

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb [hunk 2]
```diff
@@ -69,14 +69,26 @@ def set_secrets
...
 
     private
 
+    def build_selector_for_container(container)
+      if project.docker_image_building_disabled?
+        # TODO: weird that there is no opt-out here
+        [nil, container.fetch(:image)]
+      else
+        selected = container[:"samson/dockerfile"] || 'Dockerfile'
+        next if selected == 'none'
```
Comment (jonmoter) [inline plugins/kubernetes/app/models/kubernetes/template_filler.rb]: This is causing an error. Can't use next when not in a loop.

Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/template_filler.rb]: just need your time for the review, I can debug travis ;)

File: lib/samson/build_finder.rb [hunk 3]
```diff
@@ -26,51 +26,56 @@ def cancelled!
...
+            all << create_build(dockerfile)
+          end
+        end
+        needed.empty? # stop the waiting when we got everything
+      end
+
+      all
+    end
+
+    def self.detect_build_by_selector!(builds, dockerfile, image, fail:)
```
Comment (ragurney) [inline lib/samson/build_finder.rb]: Would this be more clear as:
```ruby
    class << self
      def detect_build_by_selector!(builds, dockerfile, image, fail:)
        image_name = image.split('/').last.split(/[:@]/, 2).first if image

        if (matching_build = find_matching_build(builds, dockerfile, image_name))
          matching_build
        elsif fail
          raise(
            Samson::Hooks::UserError,
            "Did not find build for dockerfile #{dockerfile.inspect} or image_name #{image_name.inspect}.\n" \
            "Found builds: #{builds.map { |b| [b.dockerfile, b.image_name] }.uniq.inspect}."
          )
        end
      end

      private

      def find_matching_build(builds, dockerfile, image_name)
        builds.detect do |b|
          (image_name && b.image_name == image_name) || (dockerfile && b.dockerfile == dockerfile)
        end
      end
    end
```
?

**Comment by grosser** [comment]
kinda similar ... maybe flattening it our with a early return/raise helps,
I'll play with this a bit

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb [hunk 4]
```diff
@@ -69,14 +69,25 @@ def set_secrets
...
+        [selected, nil]
```
Comment (ragurney) [inline plugins/kubernetes/app/models/kubernetes/template_filler.rb]: Would this break if someone has their dockerfile as `Dockerfile` and specifies an image name on their build? Then they would have a build with a [nil docker file](https://github.com/zendesk/samson/blob/master/app/models/build.rb#L76) and we wouldn't be able to find a matching build.

**Comment by jonmoter** [comment]
I'd rather focus on getting the fix out and making it look cleaner later. There are devs who are blocked by this bug, and unable to deploy.

**Comment by grosser** [comment]
oh yeah ... all green without having breaking any tests, I got high hopes for this one :D

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb (hunk 4 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/template_filler.rb]: yes, builds with nil `dockerfile` column cannot be found when not using `external images` feature, but that's a separate discussion/feature ... trying to keep this refactoring only :)

**Pull request merged**